src/Lecture2.hs

```diff
@@ -292,4 +404,45 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding (Lit lit) = Lit lit
```

Comment:
It's possible to implement this function using pattern matching with fewer constructors. The idea is to rebalance the evaluation tree by floating all literals to one the subtrees (left or right) with a recursive descend.

Alternatively, you can go with a simpler approach:

1. Traverse tree and collect a value of type `([String], Int)`: list of all variable names and sum of all constants.
2. Pattern match on this value and construct the result eventually.

Reply:
I've tried moving out of my comfort zone and still pursue approach no. 1, but still no (complete joy). Though my second solution is simpler.